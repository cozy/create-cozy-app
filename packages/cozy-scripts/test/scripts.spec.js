/* eslint-env jest */

const fs = require('fs-extra')
const path = require('path')
const colorize = require('../utils/_colorize.js')
const spawn = require('cross-spawn')

const testFolder = '.tmp_test'
const rootPath = process.cwd()
const appName = 'test-app'
const testPath = path.join(rootPath, testFolder)
const appPath = path.join(testPath, appName)
const appCoveragePath = path.join(appPath, 'coverage/')
const customConfigPath = path.join(appPath, 'app.config.js')
const ownTestConfig = path.join(__dirname, 'lib', 'test.config.js')
const servicesTestConfig = path.join(__dirname, 'lib', 'services.config.js')
const strategyTestConfig = path.join(__dirname, 'lib', 'strategy.config.js')

process.on('SIGINT', () => {
  console.log()
  console.log()
  console.log(colorize.red('Kill signal detected. Graceful exit...'))
  cleanUp()
  process.exit(1)
})

jest.setTimeout(180000) // 180s timeout

function cleanUp () {
  console.log(colorize.orange('Cleaning up generated files'))
  process.chdir(rootPath)
  fs.removeSync(testPath)
  console.log('Cleaned.')
}

// List all files in a directory recursively synchronously
const excludedFiles = [
  '.DS_Store',
  '.Spotlight-V100',
  '.Trashes',
  'ehthumbs.db',
  'Thumbs.db',
  'desktop.ini'
]
var readDeepDirSync = function (dir, filelist, parentPath = '') {
  const files = fs.readdirSync(dir)
  filelist = filelist || []
  files.forEach(file => {
    const fileOrDirPath = path.join(dir, file)
    // it's a folder (exception for node_modules, we consider it
    // as simple file here to avoid useless noise in snapshots)
    if (fs.statSync(fileOrDirPath).isDirectory() && file !== 'node_modules') {
      const parent = `${parentPath}${file}/`
      filelist = readDeepDirSync(`${fileOrDirPath}/`, filelist, parent)
    } else {
      if (!excludedFiles.includes(file)) filelist.push(`${parentPath}${file}`)
    }
  })
  return filelist
}

const overrideData = {
  '<SLUG_GH>': appName,
  '<APP_SLUG>': appName,
  '<APP_NAME>': appName,
  '<USERNAME_GH>': 'foo'
}

function getConfig (options) {
  const getWebpackConfigs = require(path.join(appPath, 'node_modules', 'cozy-scripts', 'scripts', 'config.js'))
  let appConfig = getWebpackConfigs(options)
  // we replace path to avoid environment specific snapshots
  // ex: paths like `/me/test/${testFolder}/...` will be `${testFolder}/...`
  const pathReplaceRegex = new RegExp(`"\\S*/${testFolder}/${appName}`, 'g')
  return JSON.stringify(appConfig, null, 2)
    .replace(pathReplaceRegex, `"${testFolder}/${appName}`)
}

describe('App from cozy-scripts', () => {
  beforeAll(() => {
    // create the app test folder
    fs.ensureDirSync(appPath)
    process.chdir(appPath)
    fs.writeJsonSync('./package.json', {name: appName})
    // install the scripts and run it, this part is normally handlded by create-cozy-app
    const args = [
      'add',
      '--prefer-offline',
      '--exact' // always at the end
    ].concat([`cozy-scripts@file:${path.join(rootPath, 'packages', 'cozy-scripts')}`])
    spawn.sync('yarn', args, { stdio: 'inherit' })
  })

  beforeEach(() => {
    // reset NODE_ENV
    if (process.env.NODE_ENV) delete process.env.NODE_ENV
    if (process.env.COZY_SCRIPTS_DEBUG) delete process.env.COZY_SCRIPTS_DEBUG
    jest.resetModules()
    // rm coverage folder from tests if exists
    if (fs.existsSync(appCoveragePath)) {
      fs.removeSync(appCoveragePath)
    }
    // custom config test
    if (fs.existsSync(customConfigPath)) {
      fs.removeSync(customConfigPath)
    }
  })

  afterAll(() => {
    // clean up all
    cleanUp()
  })

  // Files outline
  it('should have the correct files outline', (done) => {
    const init = require(path.join(appPath, 'node_modules', 'cozy-scripts', 'scripts', 'init.js'))
    const options = { verbose: false }
    // run the initiallisation script
    init(appPath, appName, options, (e) => {
      console.log(colorize.red('The script exited for some reasons. The test failed.'))
      cleanUp()
      console.log(e)
      throw new Error(colorize.red('Scripts tests failed.'))
    }, overrideData, () => {
      console.log(colorize.orange('Asserting created app content (files outline)...'))
      const generatedElements = readDeepDirSync(appPath)
      expect(generatedElements).toMatchSnapshot()
      // we use done() here to force waiting this callback call
      done()
    })
  })

  // Output configuration
  it('should have the correct config browser:development by default', () => {
    console.log(colorize.orange('Asserting configs...'))
    expect(JSON.parse(getConfig())).toMatchSnapshot()
  })

  it('should have the correct config browser:development by default handling --debug mode', () => {
    process.env.COZY_SCRIPTS_DEBUG = 'true'
    expect(JSON.parse(getConfig())).toMatchSnapshot()
  })

  it('should handle correctly the config browser:development', () => {
    const appConfigFromParams = getConfig({
      mode: 'development',
      target: 'browser'
    })
    process.env.NODE_ENV = 'browser:development'
    const appConfigFromEnv = getConfig()
    expect(appConfigFromParams).toEqual(appConfigFromEnv)
    expect(JSON.parse(appConfigFromParams)).toMatchSnapshot()
  })

  it('should handle correctly the config browser:production', () => {
    const appConfigFromParams = getConfig({
      mode: 'production',
      target: 'browser'
    })
    process.env.NODE_ENV = 'browser:production'
    const appConfigFromEnv = getConfig()
    expect(appConfigFromParams).toEqual(appConfigFromEnv)
    expect(JSON.parse(appConfigFromParams)).toMatchSnapshot()
  })

  it('should handle correctly the config mobile:development', () => {
    const appConfigFromParams = getConfig({
      mode: 'development',
      target: 'mobile'
    })
    process.env.NODE_ENV = 'mobile:development'
    const appConfigFromEnv = getConfig()
    expect(appConfigFromParams).toEqual(appConfigFromEnv)
    expect(JSON.parse(appConfigFromParams)).toMatchSnapshot()
  })

  it('should handle correctly the config mobile:production', () => {
    const appConfigFromParams = getConfig({
      mode: 'production',
      target: 'mobile'
    })
    process.env.NODE_ENV = 'mobile:production'
    const appConfigFromEnv = getConfig()
    expect(appConfigFromParams).toEqual(appConfigFromEnv)
    expect(JSON.parse(appConfigFromParams)).toMatchSnapshot()
  })

  // Generated app tests
  it('should pass all app tests with success', () => {
    console.log(colorize.orange('Running app tests...'))
    const testScript = require(path.join(appPath, 'node_modules', 'cozy-scripts', 'scripts', 'test.js'))
    expect(() => {
      testScript({
        cliArgs: ['--verbose', '--coverage']
      })
    }).not.toThrow()
  })

  // the --help option should always be a working option of the publish CLI
  it('should work correctly with publish --help option', () => {
    console.log(colorize.orange('Running publish --help...'))
    const publishScript = require(path.join(appPath, 'node_modules', 'cozy-scripts', 'scripts', 'publish.js'))
    expect(() => {
      publishScript({
        cliArgs: ['--help']
      })
    }).not.toThrow()
  })

  // Custom app.config.js
  it('should use the custom app config if an `app.config.js` exists in the app directory', () => {
    fs.copySync(ownTestConfig, customConfigPath)
    expect(JSON.parse(getConfig())).toMatchSnapshot()
  })

  it('should return two separated configs if services config is used (also test app.config.js usage)', () => {
    fs.copySync(servicesTestConfig, customConfigPath)
    fs.ensureDirSync(path.join(appPath, 'src', 'targets', 'services'))
    fs.writeFileSync(path.join(appPath, 'src', 'targets', 'services', 'testservice.js'), '')
    expect(JSON.parse(getConfig())).toMatchSnapshot()
  })

  it('should correctly use the webpack-merge strategy if a __mergeStrategy property is found', () => {
    fs.copySync(strategyTestConfig, customConfigPath)
    expect(JSON.parse(getConfig())).toMatchSnapshot()
  })

  // Webpack running
  it('should run webpack.run correctly with build script', (done) => {
    console.log(colorize.orange('Testing cozy-scripts build script...'))
    // should be NODE_ENV = 'browser:production' by default here
    const build = require(path.join(appPath, 'node_modules', 'cozy-scripts', 'scripts', 'build.js'))
    expect(() => build({}, done)).not.toThrow()
    expect(process.env.NODE_ENV).toBe('browser:production')
  })

  // should be always at the end (due to cleanUp usage)
  it('should run webpack.watch correctly with watch script', (done) => {
    console.log(colorize.orange('Testing cozy-scripts watch script...'))
    // should be NODE_ENV = 'browser:development' by default here
    const watch = require(path.join(appPath, 'node_modules', 'cozy-scripts', 'scripts', 'watch.js'))
    expect(() => watch({}, multiWatching => {
      multiWatching.close(() => {
        let isClosed = true
        for (const watching of multiWatching.watchings) {
          isClosed = isClosed && watching.closed
        }
        if (isClosed) {
          console.log(colorize.orange('Watch script closed correctly.'))
        } else {
          cleanUp()
          throw new Error(colorize.red('The watch script may not have been closed correctly'))
        }
        done()
      })
    })).not.toThrow()
    expect(process.env.NODE_ENV).toBe('browser:development')
  })
})
