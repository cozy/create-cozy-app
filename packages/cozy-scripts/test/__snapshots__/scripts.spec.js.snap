// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App from cozy-scripts should correctly use the webpack-merge strategy if a __mergeStrategy property is found 1`] = `
Array [
  Object {
    "module": Object {
      "rules": Array [
        Object {
          "test": Object {},
          "use": Array [
            ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/mini-css-extract-plugin/dist/loader.js",
            Object {
              "loader": "css-loader",
              "options": Object {
                "importLoaders": 1,
                "sourceMap": true,
              },
            },
            Object {
              "loader": "postcss-loader",
              "options": Object {
                "ident": "postcss",
                "sourceMap": true,
              },
            },
          ],
        },
        Object {
          "exclude": Object {},
          "test": Object {},
          "use": Array [
            ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/mini-css-extract-plugin/dist/loader.js",
            Object {
              "loader": "css-loader",
              "options": Object {
                "importLoaders": 1,
                "localIdentName": "[local]--[hash:base64:5]",
                "modules": true,
                "sourceMap": true,
              },
            },
            Object {
              "loader": "postcss-loader",
              "options": Object {
                "sourceMap": true,
              },
            },
            Object {
              "loader": "stylus-loader",
              "options": Object {
                "preferPathResolver": "webpack",
              },
            },
          ],
        },
      ],
    },
    "resolve": Object {
      "extensions": Array [
        ".styl",
      ],
    },
  },
]
`;

exports[`App from cozy-scripts should have the correct files outline 1`] = `
Array [
  ".editorconfig",
  ".eslintrc.json",
  ".github/ISSUE_TEMPLATE",
  ".github/PULL_REQUEST_TEMPLATE",
  ".gitignore",
  ".prettierrc",
  ".stylintrc",
  ".travis.yml",
  "CONTRIBUTING.md",
  "LICENSE",
  "README.md",
  "babel.config.js",
  "jest.config.js",
  "manifest.webapp",
  "node_modules",
  "package.json",
  "renovate.json",
  "src/__mocks__/fileMock.js",
  "src/__snapshots__/app.spec.js.snap",
  "src/app.spec.js",
  "src/assets/icons/icon-bullet-point.svg",
  "src/components/App.jsx",
  "src/components/HelloViews/Hello1.jsx",
  "src/components/HelloViews/Hello2.jsx",
  "src/components/Sidebar.jsx",
  "src/components/Todos/TodoAdd.jsx",
  "src/components/Todos/TodoRemoveButton.jsx",
  "src/components/Todos/TodosList.jsx",
  "src/components/Todos/__snapshots__/index.spec.js.snap",
  "src/components/Todos/__snapshots__/todoAdd.spec.jsx.snap",
  "src/components/Todos/__snapshots__/todoRemoveButton.spec.js.snap",
  "src/components/Todos/__snapshots__/todosList.spec.js.snap",
  "src/components/Todos/_mocksTodos.json",
  "src/components/Todos/index.jsx",
  "src/components/Todos/index.spec.js",
  "src/components/Todos/todoAdd.spec.jsx",
  "src/components/Todos/todoRemoveButton.spec.js",
  "src/components/Todos/todosList.spec.js",
  "src/components/__snapshots__/helloViews.spec.js.snap",
  "src/components/__snapshots__/sidebar.spec.js.snap",
  "src/components/helloViews.spec.js",
  "src/components/sidebar.spec.js",
  "src/doctypes/index.js",
  "src/doctypes/todos.js",
  "src/doctypes/todos.spec.js",
  "src/jest.setup.js",
  "src/lib/I18n.js",
  "src/locales/en.json",
  "src/styles/index.styl",
  "src/styles/todos.css",
  "src/targets/browser/index.ejs",
  "src/targets/browser/index.jsx",
  "src/targets/intents/.gitkeep",
  "src/targets/mobile/DOCS.md",
  "src/targets/mobile/config.xml",
  "src/targets/vendor/assets/icon.svg",
  "yarn.lock",
]
`;

exports[`App from cozy-scripts should return two separated configs if services config is used (also test app.config.js usage) 1`] = `
Array [
  Object {
    "bail": true,
    "devtool": "cheap-module-eval-source-map",
    "entry": Object {
      "app": Array [
        ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/@babel/polyfill/lib/index.js",
        ".tmp_test/test-app/node_modules/cozy-scripts/utils/reactExposer.js",
        ".tmp_test/test-app/src/targets/browser/index.jsx",
      ],
    },
    "externals": Array [
      "cozy",
    ],
    "mode": "development",
    "module": Object {
      "noParse": Array [
        Object {},
      ],
      "rules": Array [
        Object {
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/babel-loader/lib/index.js",
          "options": Object {
            "cacheDirectory": "node_modules/.cache/babel-loader/js",
            "presets": Array [
              Array [
                "cozy-app",
                Object {
                  "react": false,
                },
              ],
            ],
          },
          "test": Object {},
        },
        Object {
          "test": Object {},
          "use": Array [
            ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/mini-css-extract-plugin/dist/loader.js",
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/css-loader/dist/cjs.js",
              "options": Object {
                "importLoaders": 1,
              },
            },
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/postcss-loader/src/index.js",
              "options": Object {
                "ident": "postcss",
                "sourceMap": true,
              },
            },
          ],
        },
        Object {
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/file-loader/dist/cjs.js",
          "options": Object {
            "name": "[name].[ext]",
          },
          "test": Object {},
        },
        Object {
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/babel-loader/lib/index.js",
          "options": Object {
            "cacheDirectory": "node_modules/.cache/babel-loader/react",
            "plugins": Array [],
            "presets": Array [
              "cozy-app",
            ],
          },
          "test": Object {},
        },
        Object {
          "enforce": "pre",
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/eslint-loader/index.js",
          "options": Object {
            "emitWarning": true,
            "extends": Array [
              "cozy-app",
            ],
            "fix": false,
          },
          "test": Object {},
        },
        Object {
          "enforce": "pre",
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/eslint-loader/index.js",
          "options": Object {
            "emitWarning": true,
            "extends": Array [
              "cozy-app/react",
            ],
            "fix": false,
          },
          "test": Object {},
        },
        Object {
          "exclude": Object {},
          "test": Object {},
          "use": Array [
            ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/mini-css-extract-plugin/dist/loader.js",
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/css-loader/dist/cjs.js",
              "options": Object {
                "importLoaders": 1,
                "sourceMap": true,
              },
            },
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/postcss-loader/src/index.js",
              "options": Object {
                "ident": "postcss",
                "sourceMap": true,
              },
            },
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/stylus-loader/index.js",
              "options": Object {
                "preferPathResolver": "webpack",
                "use": Array [
                  null,
                ],
              },
            },
          ],
        },
        Object {
          "include": Object {},
          "test": Object {},
          "use": Array [
            ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/mini-css-extract-plugin/dist/loader.js",
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/css-loader/dist/cjs.js",
              "options": Object {
                "importLoaders": 1,
                "localIdentName": "[local]--[hash:base64:5]",
                "modules": true,
                "sourceMap": true,
              },
            },
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/postcss-loader/src/index.js",
              "options": Object {
                "ident": "postcss",
                "sourceMap": true,
              },
            },
            Object {
              "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/stylus-loader/index.js",
              "options": Object {
                "use": Array [
                  null,
                ],
              },
            },
          ],
        },
        Object {
          "include": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/svg-sprite-loader/lib/loader.js",
          "options": Object {
            "symbolId": "[name]_[hash]",
          },
          "test": Object {},
        },
        Object {
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/file-loader/dist/cjs.js",
          "options": Object {
            "name": "[name].[ext]",
            "outputPath": "img/",
            "publicPath": "/img",
          },
          "test": Object {},
        },
        Object {
          "include": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/file-loader/dist/cjs.js",
          "options": Object {
            "name": "[name].[ext]",
            "outputPath": "public/img",
            "publicPath": "/public/img",
          },
          "test": Object {},
        },
        Object {
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/json-loader/index.js",
          "test": Object {},
        },
        Object {
          "loader": "imports-loader?css=./cozy-bar.min.css",
          "test": Object {},
        },
      ],
    },
    "optimization": Object {
      "removeAvailableModules": false,
      "removeEmptyChunks": false,
      "splitChunks": Object {
        "cacheGroups": Object {
          "vendors": Object {
            "name": "vendors",
            "test": Object {},
          },
        },
      },
    },
    "output": Object {
      "filename": "[name]/test-app.js",
      "path": ".tmp_test/test-app/build",
      "pathinfo": false,
    },
    "plugins": Array [
      Object {
        "options": Object {
          "chunkFilename": "[name]/test-app.css",
          "filename": "[name]/test-app.css",
        },
      },
      Object {
        "plugins": Array [
          null,
          null,
          null,
        ],
        "test": Object {},
      },
      Object {
        "definitions": Object {
          "__ENABLED_FLAGS__": "[]",
        },
      },
      Object {
        "config": Object {
          "plainSprite": false,
          "spriteAttrs": Object {},
        },
        "rules": Object {},
        "svgCompiler": Object {
          "config": Object {
            "rules": Array [
              Object {
                "test": Object {},
                "value": "sprite.svg",
              },
            ],
          },
          "rules": Object {
            "rules": Array [
              Object {
                "test": Object {},
                "value": "sprite.svg",
              },
            ],
          },
          "symbols": Array [],
        },
      },
      Object {
        "config": Object {
          "plainSprite": false,
          "spriteAttrs": Object {},
        },
        "rules": Object {},
        "svgCompiler": Object {
          "config": Object {
            "rules": Array [
              Object {
                "test": Object {},
                "value": "sprite.svg",
              },
            ],
          },
          "rules": Object {
            "rules": Array [
              Object {
                "test": Object {},
                "value": "sprite.svg",
              },
            ],
          },
          "symbols": Array [],
        },
      },
      Object {},
      Object {},
      Object {
        "modulesCount": 500,
        "profile": false,
        "showActiveModules": true,
        "showEntries": false,
        "showModules": true,
      },
      Object {
        "options": Object {
          "cache": true,
          "chunks": Array [
            "vendors",
            "app",
          ],
          "chunksSortMode": "auto",
          "compile": true,
          "excludeChunks": Array [],
          "favicon": false,
          "filename": "index.html",
          "hash": false,
          "inject": false,
          "meta": Object {},
          "minify": Object {
            "collapseWhitespace": true,
          },
          "showErrors": true,
          "template": ".tmp_test/test-app/src/targets/browser/index.ejs",
          "title": "test-app",
          "xhtml": false,
        },
      },
      Object {
        "options": Object {
          "async": Object {
            "test": Array [],
          },
          "custom": Array [],
          "defaultAttribute": "defer",
          "defer": Object {
            "test": Array [],
          },
          "inline": Object {
            "test": Array [],
          },
          "module": Object {
            "test": Array [],
          },
          "prefetch": Object {
            "chunks": "initial",
            "test": Array [],
          },
          "preload": Object {
            "chunks": "initial",
            "test": Array [],
          },
          "removeInlinedAssets": true,
          "sync": Object {
            "test": Array [],
          },
        },
      },
      Object {
        "definitions": Object {
          "__TARGET__": "\\"browser\\"",
        },
      },
      Object {
        "definitions": Object {
          "__DEVELOPMENT__": true,
        },
      },
      Object {
        "definitions": Object {
          "__STACK_ASSETS__": false,
        },
      },
      Object {
        "definitions": Object {
          "cozy.bar": "cozy-bar/dist/cozy-bar.min.js",
          "cozy.client": "cozy-client-js/dist/cozy-client.min.js",
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".js",
        ".json",
        ".css",
        ".jsx",
        ".styl",
      ],
      "modules": Array [
        ".tmp_test/test-app/src",
        ".tmp_test/test-app/node_modules",
      ],
      "symlinks": false,
    },
  },
  Object {
    "bail": true,
    "devtool": false,
    "entry": Object {
      "testservice": ".tmp_test/test-app/src/targets/services/testservice.js",
    },
    "externals": Array [],
    "mode": "development",
    "module": Object {
      "rules": Array [
        Object {
          "enforce": "pre",
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/eslint-loader/index.js",
          "options": Object {
            "emitWarning": true,
            "extends": Array [
              "cozy-app",
            ],
            "fix": false,
          },
          "test": Object {},
        },
        Object {
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/babel-loader/lib/index.js",
          "options": Object {
            "babelrc": false,
            "cacheDirectory": "node_modules/.cache/babel-loader/node",
            "presets": Array [
              Array [
                "cozy-app",
                Object {
                  "node": true,
                  "react": false,
                },
              ],
            ],
          },
          "test": Object {},
        },
        Object {
          "enforce": "pre",
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/eslint-loader/index.js",
          "options": Object {
            "emitWarning": true,
            "extends": Array [
              "cozy-app",
            ],
            "fix": false,
          },
          "test": Object {},
        },
        Object {
          "exclude": Object {},
          "loader": ".tmp_test/test-app/node_modules/cozy-scripts/node_modules/babel-loader/lib/index.js",
          "options": Object {
            "babelrc": false,
            "cacheDirectory": "node_modules/.cache/babel-loader/node",
            "presets": Array [
              Array [
                "cozy-app",
                Object {
                  "node": true,
                  "react": false,
                },
              ],
            ],
          },
          "test": Object {},
        },
      ],
    },
    "optimization": Object {},
    "output": Object {
      "filename": "[name]/test-app.js",
      "path": ".tmp_test/test-app/build/services",
    },
    "plugins": Array [
      Object {
        "definitions": Object {
          "__TARGET__": "\\"services\\"",
        },
      },
      Object {
        "definitions": Object {
          "__TARGET__": "\\"services\\"",
        },
      },
    ],
    "resolve": Object {
      "extensions": Array [
        ".js",
        ".json",
        ".css",
        ".jsx",
        ".styl",
      ],
      "modules": Array [
        ".tmp_test/test-app/src",
        ".tmp_test/test-app/node_modules",
      ],
      "symlinks": false,
    },
    "target": "node",
  },
]
`;
